Template for packaging spinalcordtoolbox datasets in pip.

kousu opened the pull request.

Commit:
Template for packaging spinalcordtoolbox datasets in pip.

kousu commented:
@joshuacwnewton

joshuacwnewton commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,24 @@
+      run: |
+          python -m pip install --upgrade pip
+          pip install build
+    - name: Build
+      run: |
+          python -m build --wheel --sdist
```
Could we take this a step further by then installing the built package, then importing it (e.g. with `python -c`)? That might be a more thorough test to see if everything is working as expected.

kousu replied:
Clever.

But given the mess in https://github.com/sct-data/spinalcordtoolbox-data-template/pull/1/files#r608778185 there's a lot of combinations to test. And how would you test every subdirectory?

Maybe we could run `find src/ -type -f ! -name "*.py"` and make sure each of those paths are importable via `importlib.resources`.

joshuacwnewton replied:
Couldn't we use `find_namespace_packages`, since that's what `setup.py` is using, too?

joshuacwnewton replied:
This is really rough pseudocode, but:

```
pip install *.whl
python -c "import setuptools, importlib; [importlib.import(package) for package in setuptools.find_namespace_packages(<site-packages-dir>)]"
```

kousu replied:
oh that is very clever. i'll see about working that in.

kousu replied:
`importlib.metadata` is in the standard library so I'd like to try to use that if possible. E.g. I figured out you can do:

```
>>> pprint(importlib.metadata.distribution('spinalcordtoolbox-data-catscatscats').files)
[PackagePath('spinalcordtoolbox/data/dataset/__init__.py'),
 PackagePath('spinalcordtoolbox/data/dataset/__pycache__/__init__.cpython-39.pyc'),
 PackagePath('spinalcordtoolbox/data/dataset/cattiestcats.txt'),
 PackagePath('spinalcordtoolbox_data_catscatscats-0.1.dev9+g941c563.d20210713.dist-info/INSTALLER'),
 PackagePath('spinalcordtoolbox_data_catscatscats-0.1.dev9+g941c563.d20210713.dist-info/METADATA'),
 PackagePath('spinalcordtoolbox_data_catscatscats-0.1.dev9+g941c563.d20210713.dist-info/RECORD'),
 PackagePath('spinalcordtoolbox_data_catscatscats-0.1.dev9+g941c563.d20210713.dist-info/REQUESTED'),
 PackagePath('spinalcordtoolbox_data_catscatscats-0.1.dev9+g941c563.d20210713.dist-info/WHEEL'),
 PackagePath('spinalcordtoolbox_data_catscatscats-0.1.dev9+g941c563.d20210713.dist-info/direct_url.json'),
 PackagePath('spinalcordtoolbox_data_catscatscats-0.1.dev9+g941c563.d20210713.dist-info/top_level.txt')]
```

kousu replied:
some mad bad wizardry

```
[kousu@requiem data-template]$ diff --label src  <((cd src; find spinalcordtoolbox -type f) | sort)  --label installed <(python -c 'import importlib.metadata; [print(e) for e in importlib.metadata.distribution("spinalcordtoolbox-data-catscatscats").files]' | grep -v '.dist-info' | grep -v "*.pyc" | sort)
--- src
+++ installed
@@ -1,2 +1,3 @@
 spinalcordtoolbox/data/dataset/cattiestcats.txt
 spinalcordtoolbox/data/dataset/__init__.py
+spinalcordtoolbox/data/dataset/__pycache__/__init__.cpython-39.pyc
```

kousu replied:
On the other hand this is pretty complicated, and the fact that we're using `setuptools_scm` + doing releases *via* github which forces files to be committed first means it's pretty unlikely we can make a broken package.

And if we do, people can open an issue.

joshuacwnewton commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,44 @@
+name: Publish
+
+on:
+  # publish from the Releases page:
+  release:
+    types: [published]
```
Dumb question: Why doesn't this create an infinite loop with the "Publish to GitHub" step?

joshuacwnewton replied:
Oh, I think I understand... Does this workflow file add the sdist/wheel to an existing release (rather than triggering a new release entirely)?

kousu replied:
That's right. It wasn't clear to me from [the docs for `softprops/action-gh-release@v1`](https://github.com/marketplace/actions/gh-release) immediately either, but I tried it and double checked by reading its code: since each release goes with a specific tag, it creates the release for that tag or reuses the existing release if it's already there.

And btw I chose `published` over `created` because we had this problem last week in ivadomed where there was a scrap release made by accident because a release got Drafted, the workflow ran, but decided not to finish because it wasn't Published, and then there was no obvious way to rerun it. I think only publishing on Published should avoid that.

joshuacwnewton commented on README.md:
```diff
@@ -0,0 +1,41 @@
+# <dataset> template
+
+Part of [`spinalcordtoolbox`](https://github.com/neuropoly/spinalcordtoolbox).
```
Could we expand this section to include "How to use the pip package" instructions? 

e.g. "Once this package is installed via pip, the files will be placed in `site-packages/...`. You can access this directory in your code by calling `importlib...`"

joshuacwnewton commented on README.md:
```diff
@@ -0,0 +1,41 @@
+    2. Fill in a version tag. We use [this versioning policy](TODO)
+    3. Click "Publish Release"
+    4. Wait a few minutes;
+    5. Monitor the progress at https://github.com/neuropoly/spinalcordtoolbox-data-${dataset_name}/actions/workflows/release.yml
+    6. The release should appear on https://github.com/neuropoly/spinalcordtoolbox-data-${dataset_name}/releases
```
`neuropoly/` -> `sct-data/`?

kousu replied:
OUT OF DATE

joshuacwnewton replied:
`spinalcordtoolbox` :D

kousu replied:
`spinalcordtoolbox/`

joshuacwnewton commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+  #license_files=[ ... ] # TODO?
+
+  packages=find_namespace_packages('src/'),
+  package_dir={"":"src/"},
```
This is awesome! I didn't realize the package directory could be changed.

kousu replied:
I got this from https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure.

I consider it a wart. But we need it to go with the previous line.

joshuacwnewton commented on src/spinalcordtoolbox/data/dataset/__init__.py:
```diff
@@ -0,0 +1,2 @@
+# empty __init__.py to enable importlib.resources
+# see < TODO >
```
Which TODO is this referring to?

kousu replied:
There's another terrible python wart I ran into when setting this up that I don't know how or where to explain.


The basic issue is, `sys.path` behaves like `$PATH` or like `overlayfs`: it implies a virtual filesystem made of the **union** of all the folders listed in it. But python only pretends to support this and needs lots of workarounds for it. Recently, https://www.python.org/dev/peps/pep-0420/ dropped the necessity of `__init__.py` to create a package, in order to make it easier for this virtual union filesystem thing to work out; otherwise, every package that did what I did here, share a folder (`spinalcordtoolbox/data`), would be have to ship a `__init__.py` and would therefore be fighting with each other. So, A+ to that; but for some reason in the process they never applied the same thinking to non-python files. There's even this comment: https://github.com/python/cpython/blob/30e507dff465a31901d87df791a2bac40dc88530/Lib/importlib/resources.py#L243-L247

> By definition, namespace packages cannot have resources

So what this means is this terrible confusing situation where:

- if you `pip install` everything, everything ends up in the same physical directories and things make sense
    - and you could distribute an `__init__.py` everywhere like you *used* to before PEP420, and just be careful not to create any conflicts, and then everything would work fine
- if you `pip install -e` one or more of the packages, then you're in this virtual situation where the filesystem you see via `import` and `importlib` is not the same as on disk but you can access python code but you can't access the others.
    - and if you try to distribute `__init__.py`s in this situation, you can't access files from the other apps
        - specifically, [`spinalcordtoolbox/__init__.py`](https://github.com/neuropoly/spinalcordtoolbox/blob/91e9764c7bce79972e445bb61c247ec2f00e49d2/spinalcordtoolbox/__init__.py) shadows the entire `spinalcordtoolbox/data/` subdirectory, so none of these packages will work until we remove that, *but only if you've `pip install -e spinalcordtoolbox`'d*. It will work if you `pip install spinalcordtoolbox`. ![](http://www.sherv.net/cm/emo/angry/pulling-hair.gif)

BUT they literally just fixed this a month ago: https://github.com/python/cpython/pull/24670. Praise the python maintainers. So maybe that link is what can go in the TODO.

kousu replied:
This tiny TODO took me like 4 hours to figure out. It definitely is where the bulk of my time working on this went and I don't know how to explain what I was doing because it's such a corner case.

And it has implications for the rest of sct. It means we need to delete all the __init__.pys from it. I checked and mostly they're empty anyway, but

* https://github.com/neuropoly/spinalcordtoolbox/blob/91e9764c7bce79972e445bb61c247ec2f00e49d2/spinalcordtoolbox/__init__.py
* https://github.com/neuropoly/spinalcordtoolbox/blob/91e9764c7bce79972e445bb61c247ec2f00e49d2/spinalcordtoolbox/utils/__init__.py

aren't totally.

But (until python3.10 I guess, when we finally get https://github.com/python/cpython/pull/24670) we'll have to keep these empty `__init__.py`s *in the data folders* but not in the *code* folders. Seems pretty fragile.

Or it looks like https://github.com/python/importlib_resources/ is still being maintained as a backport; I thought it was just for python3.6 but maybe it even covers newer versions?

joshuacwnewton replied:
> And it has implications for the rest of sct. It means we need to delete all the **init**.pys from it. I checked and mostly they're empty anyway, but

Is this true? PEP420 [only applies to namespace packages](https://www.python.org/dev/peps/pep-0420/#specification). 

> Regular packages will continue to have an `__init__.py` and will reside in a single directory.
>
> Namespace packages cannot contain an `__init__.py`. 

Namespace packages are defined as [packages that are split up into multiple distributions](https://packaging.python.org/guides/packaging-namespace-packages/). I don't think this describes the subpackages inside `spinalcordtoolbox`, does it? (Since we distribute everything as a single package, we're more like the first directory tree from [that link](https://packaging.python.org/guides/packaging-namespace-packages/), rather than the second.) 

So, I think we should be able to leave the `__init__.py` files alone.

kousu replied:
I almost want to throw out `importlib.resources` entirely and just go back to using `__file__`. If we put:

```
# __init__.py

def contents():
    dir = os.path.dirname(__file__)
    return [os.path.join(dir, file) for file in os.listdir(dir)]
```

in every `__init__.py` we would have the same results for both

```
import spinalcordtoolbox.data.something

importlib.resources.contents(spinalcordtoolbox.data.something)
spinalcordtoolbox.data.something.contents()
```

 . It would mean we wouldn't be able to use virtual namespaces, so we couldn't make a multi-part `spinalcordtoolbox-data-something` package like python3.10 apparently can, but why would we want to?

...maybe we would want to do that to deal with splitting up data to fit online? Like a multi-part .rar or  a replacement for `ghsplit`?

joshuacwnewton replied:
> import spinalcordtoolbox.data.something
> 
> importlib.resources.contents(spinalcordtoolbox.data.something)
> spinalcordtoolbox.data.something.contents()

This does feel a lot more readable. I was worried about `importlib` not being very self-evident. I also like how this can be handled on the template level.

Looking at the current usage, though (e.g. [`sct_deepseg`](https://github.com/neuropoly/spinalcordtoolbox/blob/master/spinalcordtoolbox/deepseg/models.py) with its use of `folder`) I feel like exposing the base directory might line up more with how it parses for files.

kousu replied:
Okay actually I think this provides a solution:

* https://github.com/python/importlib_resources/issues/68 -> https://github.com/python/importlib_resources/pull/196 fixed this in `importlib_resources`, the backport
* https://bugs.python.org/issue42129 -> https://github.com/python/cpython/pull/24670 fixes this in `importlib.resources`

So we can just:

```
install_requires=[
  ...
  importlib_resources>=4
  ...
],
```

and

```
import importlib_resources as resources
# from importlib import resources # uncomment when https://github.com/python/cpython/pull/24670 is available in python3.10
```

and delete all `__init__.py`s and tell people to stop making new ones.

kousu replied:
I still kind of like just using `__file__`. But `resources` is the Way Of The Future so maybe we should go with that.

joshuacwnewton commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,44 @@
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Version (e.g. 2.0.3)'
```
It might be good to update this with a date-based version. (Though, I guess the versioning policy is currently a TODO.)

kousu replied:
oh snap, good catch.

[PEP440 explicitly supports dates](https://www.python.org/dev/peps/pep-0440/#support-for-date-based-version-identifiers), but under duress, I think. I think it would be good to move away from them since our software doesn't use them. Something to discuss with the rest of the team maybe?

kousu replied:
I'm not totally sure this works like I think. When you click it you get this box:

![Screenshot from 2021-04-11 03-51-11](https://user-images.githubusercontent.com/987487/114296397-3d97a200-9a79-11eb-8f96-15f3cd6a43ed.png)

If you change the branch will it tag *that one*? We need to confirm which way this goes that before committing to it.

joshuacwnewton replied:
Oh, that's interesting. I'm not sure!

(When I was testing `workflow_dispatch` for the changelog PR, my personal repo only had a single branch, so I haven't run into this yet. But, the same dropdown is present on SCT, too, so I'll have to face this as well.)

joshuacwnewton commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,44 @@
+      uses: pypa/gh-action-pypi-publish@release/v1
+      with:
+        user: __token__
+        #password: ${{ secrets.PYPI_PASSWORD }}
+        # DEBUG:
+        password: ${{ secrets.TEST_PYPI_API_TOKEN }}
+        repository_url: https://test.pypi.org/legacy/
```
This is waiting on the PYPI password requests from yesterday's meeting, right?

kousu replied:
Julien took care of that for me. They're now `secrets.PYPI_SCT_DATA` and `secrets.TEST_PYPI_SCT_DATA`.

I had a fun idea: add a flag to the `workflow_dispatch` input letting you choose TestPyPI.

joshuacwnewton commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,44 @@
+    - name: Build
+      run: |
+          python -m build --wheel --sdist
+    ### TODO: can the uploads be parallelized?
```
I [don't think parallelization is currently possible](https://github.community/t/steps-in-parallel/16343) with `steps:`. :cry:

kousu replied:
Yeah... it is with jobs though?

I managed to make it work, but only [with](https://github.com/kousu/ANTs/blob/329ef595b00538abce365453eaec99371c2ec394/.github/workflows/ccpp.yml#L163)

>      uses: actions/upload-artifact@v2-preview

and

>     - uses: actions/download-artifact@v1

and a lot of copy-pasting.

joshuacwnewton commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+      'Source': 'https://github.com/sct-data/<dataset>',
+      #'Documentation': '',
+  },
+  #license='CC-BY-NC', ??
+  #license_files=[ ... ] # TODO?
```
We settled on CC in the meeting, right?

kousu replied:
I'm not totally sure.

@jcohenadad do you want CC BY-NC-SA 4.0 or CC BY-SA 4.0 or CC BY 4.0 ?

Also licensing is a thorny legal question. We can't just slap a license on a work; at a minimum we need to contact everyone in the git log and get their thoughts, otherwise the work is proprietary to everyone who has ever worked on it, by default.

I guess for the template we can fill in an aspirational one, but for the actual datasets it's a lot trickier.

jcohenadad replied:
I'm OK with `CC BY 4.0`

joshuacwnewton replied:
```suggestion
  #license='CC BY 4.0',
  #license_files=[ LICENSE.txt ]
```

joshuacwnewton replied:
TODO: Fetch CC and put as file

joshuacwnewton commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+  url='https://spinalcordtoolbox.com/',
+  project_urls={
+      'Source': 'https://github.com/sct-data/<dataset>',
+      #'Documentation': '',
```
Could we maybe point this to `spinalcordtoolbox.com` as a sort of blanket link for any future documentation on our datasets/models/etc.?

We could even have a comment here saying "# FIXME: Update with dataset-specific link if applicable", but at the very least this way we would have a safe fallback link.

kousu replied:
Nice that's the sort of feedback I was hoping for

joshuacwnewton commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+  # with setuptools_scm, means it includes non-python files if they're under git
+  include_package_data=True,
+
+  # with setuptools_scm, get the version out of the most recent git tag.
+  # the tags must be formatted as semver.
+  use_scm_version=True,
```
This can also be mirrored in `pyproject.toml`, right? From https://pypi.org/project/setuptools-scm/: 

> To enable version inference, add this section to your pyproject.toml:
> 
> ```toml
> # pyproject.toml
> [tool.setuptools_scm]
> ```
> 
> Including this section is comparable to supplying use_scm_version=True in setup.py. Additionally, include arbitrary keyword arguments in that section to be supplied to get_version(). For example:
> 
> ```toml
> # pyproject.toml
> [tool.setuptools_scm]
> write_to = "pkg/version.py"
> ```

kousu replied:
Yes... but I am rebelling against pyproject.toml along with 70% of python.

I gave a shot to putting everything in pyproject.toml but it doesn't work! It doesn't support editable installs, and doesn't support ...uh I think `packages=find_namespace_packages('src/')`?

So I gave up and moved everything back to setup.py. I would rather keep everything uniquely in setup.py until the day pyproject.toml can fully replace it in one step. I wouldn't even have it except that, at least with the latest pip, with it `pip install .` produces a .dist-info directory, but `pip install -e .` produces a .egg-info directory.

Maybe I should put a comment about this inside pyproject.toml.

joshuacwnewton replied:
In that case, since the only content inside `pyproject.toml` is already duplicated in `setup.py`, couldn't we avoid having a `pyproject.toml` at all?

joshuacwnewton commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+
+# workaround a bug introduced by pyproject.toml
+# https://github.com/pypa/pip/issues/7953#issuecomment-645133255
+import site, sys; site.ENABLE_USER_SITE = True
```
Is `sys` used here?

kousu replied:
Yes. There's some dark magic I believe.

kousu replied:
Probably not

joshuacwnewton commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+  #license='CC-BY-NC', ??
+  #license_files=[ ... ] # TODO?
+
+  packages=find_namespace_packages('src/'),
```
We need to use `find_namespace_packages` because each data package is nested in subdirectories, right? Maybe a comment could be good here.

kousu replied:
yeah this is annoying. I wish python's packaging rules weren't so arcane. I wish they were just: "here's some files, use them".

joshuacwnewton replied:
```suggestion
  # This is weird but it makes it work
  packages=find_namespace_packages('src/'),
```

joshuacwnewton commented on setup.py:
```diff
@@ -0,0 +1,44 @@
+from setuptools import setup, find_packages, find_namespace_packages
```
`find_packages` is unused here, I think.

kousu replied:
:+1:

joshuacwnewton commented on README.md:
````diff
@@ -0,0 +1,41 @@
+    6. The release should appear on https://github.com/neuropoly/spinalcordtoolbox-data-${dataset_name}/releases
+       with the .tar.gz (sdist) and .whl (wheel) formats attached momentarily.
+
+### Troubleshooting
+
+You can test the repo locally with
+
+```
+pip install build &&
+python -m build --wheel --sdist &&
+pip install dist/*.whl
+```
+
+This should give you enough clues hopefully to track down any problems.
````
Might be good to mention `build.yml` here as well, if we duplicate these steps in that workflow file.

joshuacwnewton commented on README.md:
````diff
@@ -0,0 +1,41 @@
+    git push
+    ```
+
+2. Go to https://github.com/neuropoly/spinalcordtoolbox-data-${dataset_name}/releases
+
+    1. Click "Draft Release"
+    2. Fill in a version tag. We use [this versioning policy](TODO)
+    3. Click "Publish Release"
+    4. Wait a few minutes;
+    5. Monitor the progress at https://github.com/neuropoly/spinalcordtoolbox-data-${dataset_name}/actions/workflows/release.yml
+    6. The release should appear on https://github.com/neuropoly/spinalcordtoolbox-data-${dataset_name}/releases
````
Right now, it gets removed when following the steps to edit `README.md`. But, this seems like it's worth keeping in the child repos, too. 

So, maybe this should go into its own "Usage" section. (e.g. split into `How to use this repo template` (which gets removed) and `How to use {dataset_name}` (which stays))

joshuacwnewton reviewed (changes requested):
I'm wondering if we should address all of the existing TODOs/comments before making any new repos with this. :thinking:

Commit:
Update README.md

feedback:
* https://github.com/spinalcordtoolbox/spinalcordtoolbox-data-template/pull/1/files#r608804968

kousu committed:
Add files via upload

Commit:
Yep

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update setup.py

joshuacwnewton committed:
`setup.py`: Add link to front page for 'documentation'

Commit:
Update setup.py

Commit:
Create LICENSE.txt

Commit:
Update setup.py

taowa committed:
add a missing comma to setup.py

taowa committed:
replace the name string during builds for testing

taowa commented:
yay!

taowa commented:
So that last commit is definitely kind of sinful. But I think that we should be doing /something/ to validate changes to the template, lest we change it in ways that break other things.

joshuacwnewton commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,25 @@
+          pip install build
+    - name: Build
+      run: |
+          perl -pi -e 's/\${dataset_name}/sample-dataset/' setup.py
```
IIUC, this is a find and replace? Just out of curiosity, why `perl -pi -e`?

taowa commented:
> On Jul 30, 2021, at 14:36, Joshua Newton ***@***.***> wrote:
> +          perl -pi -e 's/\${dataset_name}/sample-dataset/' setup.py
> 
> IIUC, this is a find and replace?

That’s right :).

> Just out of curiosity, why perl -pi -e? (I tried googling and found this SO answer, but I still don't entirely understand the purpose.)

Largely to avoid differences between the BSD/OSX sed and the GNU sed. It could just as well be `sed -i ’s/[…]/‘`, and I’m happy to change it to that, but that doesn’t play well with the OSX sed and I wanted to (lazily) test it out before pushing.

Taowa

taowa committed:
replace google group email with a forwarder @neuropoly.org

taowa commented on README.md:
```diff
@@ -0,0 +1,66 @@
+2. Go to https://github.com/spinalcordtoolbox/data-${dataset_name}/releases
+
+    1. Click "Draft Release"
+    2. Fill in a version tag. We use [this versioning policy](TODO)
```
I'm partial to versioning as `1`, `2`, `3`, ... . It's dead easy, compliant with PEP 440 (`[N!]N(.N)*[{a|b|rc}N][.postN][.devN]`), and I don't really see how we'd end up with "major" and "minor" versions of our datasets. I suppose that if we want to be fancy, we can also support appending `.devN`.

joshuacwnewton committed:
`README.md`: Fix numbering

joshuacwnewton committed:
`README.md`: Improve readability of CLI steps

joshuacwnewton committed:
`README.md`: Split "Remove this section" step into its own last step

joshuacwnewton committed:
`README.md`: Add clearer git commit messages

joshuacwnewton committed:
`README.md`: Add clarifying indentation

joshuacwnewton committed:
`README.md`: Use date-based release versioning

This wasn't formally agreed upon, however it *is* the de facto default in the NeuroPoly lab, so
I wanted to start out by using what we already know and use. We can discuss changing this in the future.

joshuacwnewton committed:
`README.md`: Fixup `data-${dataset_name}` instances

joshuacwnewton committed:
`README.md`: Use `perl -pi -e` in README.md, too

joshuacwnewton committed:
`README.md`: Swap steps 1 and 2 in metadata

This allows the find and replace step to update the following git step

joshuacwnewton committed:
`README.md`: Fix typos with `$dataset-name`/`$dataset_name`

joshuacwnewton committed:
`README.md`: Improve readability of the `Troubleshooting` section

joshuacwnewton committed:
`README.md`: Add missing `.git`

joshuacwnewton commented:
I've pushed commits to this branch but they aren't showing up on this PR. I'm going to try closing and re-opening to see if it fixes anything.

joshuacwnewton committed:
`release.yml`: Remove "parallelized" TODO

Answer to TODO: No, steps cannot be parallelized.

However, *jobs* can run in parallel. But, to do so would require uploading
 files as artifacts, then spinning up entirely new VM instances, then downloading the artifacts.

 And, at that point, the setup of these VMs would probably take longer
 and use more resources than just running the steps sequentially.

joshuacwnewton committed:
`release.yml`: Update PYPI_API_TOKEN

The existing values were copied over from ivadomed. Here, in SCT,
I've added a temporary new token (until we can sort out NeuroPoly's
PyPI accounts).

joshuacwnewton committed:
`release.yml`: Test publishing on PRs for `ng/template`

joshuacwnewton committed:
`release.yml`: Remove `branches: ng/template` from test

I think this is meant to specify the target branch (i.e. `trunk`,
not the branch used for the PR.)

joshuacwnewton committed:
`release.yml`: Add missing perl -pi -e

joshuacwnewton commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,49 @@
+    - name: Replace dummy values in `setup.py` (only for `data-template`)
+      if: ${{ github.event.repository.name }} == "data-template"
+      run: perl -pi -e 's/\${dataset_name}/template/' setup.py
+    - name: Build
+      run: |
+          python -m build --wheel --sdist
+    - name: Publish to Github
+      uses: softprops/action-gh-release@v1
+      with:
+        files: 'dist/*'
+        fail_on_unmatched_files: true
+        tag_name: 2.0.3  #  ${{ github.event.inputs.version }} # in the workflow_dispatch case, make a new tag from the given input; in the published release case, this will be empty and will fall back to updating that release.
```
Currently, in our `setup.py` we've set the option `use_scm_version=True`, which will use the most recent **git tag** as the _package_ version number.

There's a subtle bug here, though: For the `workflow_dispatch` case, we create a tag in the Publish step. But, the publish step happens _after_ the build step, so the chosen tag won't be present for `python -m build` to pick up.

I think this means we have to try creating a tag prior to the build step?

joshuacwnewton replied:
Specifically, this manifests in this PR as:

```
ERROR    HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
         '0.1.dev1+g169fe48.d20220603' is an invalid value for Version. Error:  
         Can't use PEP 440 local versions. See                                  
        [ https://packaging.python.org/specifications/core](https://packaging.python.org/specifications/core-metadata)-metadata for m
         information.
```

The package is being built with `0.1.dev1`, when we want it to be built using the dummy example tag name I chose (`2.0.3`).

joshuacwnewton replied:
Ahh! Fascinating! `use_scm_version=True` does _not_ do what I thought it did ([CI run](https://github.com/spinalcordtoolbox/data-template/runs/6729243488?check_suite_focus=true)):

```
ERROR    HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
         '2.0.4.dev0+g5929d76.d20220603' is an invalid value for Version. Error:
         Can't use PEP 440 local versions. See                                  
        [ https://packaging.python.org/specifications/core](https://packaging.python.org/specifications/core-metadata)-metadata for m
         information.
```

Setting the tag `2.0.3` prior to release will actually auto-generate a package with version number `2.0.4.dev0`!

My best guess is that `use_scm_version=True` is intended for the workflow described in the [README.md of the `pypi-publish` action](https://github.com/pypa/gh-action-pypi-publish#advanced-release-management):

> For example, you could implement a parallel workflow that pushes every commit to TestPyPI or your own index server, like devpi. For this, you'd need to (1) specify a custom repository_url value and (2) generate a unique version number for each upload so that they'd not create a conflict. The latter is possible if you use setuptools_scm package but you could also invent your own solution based on the distance to the latest tagged commit.

In other words, `setuptools_scm` is meant for rapidly generating unique version numbers during development (e.g. `dev0` -> `dev1` -> `dev2`), rather than generating the version number for stable releases.

This means that we should probably stop using `setuptools_scm` for building stable releases! (Though, we could optionally continue to use it to automatically build test/dev releases, but that's not really what `workflow_dispatch` is for.)

joshuacwnewton replied:
Right! So, performing a find and replace on the `version="dev'` string (7394563) gets us our desired result:

* https://test.pypi.org/project/spinalcordtoolbox-data-template/2.0.3/
* https://github.com/spinalcordtoolbox/data-template/releases/tag/2.0.3

kousu replied:
That doesn't make sense to me. I've used it lots of times and its always behaved. Take a look at https://github.com/ufoym/imbalanced-dataset-sampler/actions/runs/2371746877 to see for yourself.

If you want to do a test deploy you should use just tag it 'v1.0.3rc0' or 'v1.0.3rc1' -- that'll get tagged as a test build too -- particularly we should include this line

https://github.com/ufoym/imbalanced-dataset-sampler/actions/runs/2371746877

I'm on my phone atm. I'll need some time to look deeper about why setuptools-scm isn't working. But I really think we should use it, it really improves the workflow.

joshuacwnewton replied:
> I'll need some time to look deeper about why setuptools-scm isn't working. 

This seemed to be intended/working behavior for `setuptools-scm` to me? From [`pypa/setuptools_scm`](https://github.com/pypa/setuptools_scm):

> Default versioning scheme
> 
> In the standard configuration setuptools_scm takes a look at three things:
> 
>  * latest tag (with a version number)
>  * the distance to this tag (e.g. number of revisions since latest tag)
>  * workdir state (e.g. uncommitted changes since latest tag)
> 
> and uses roughly the following logic to render the version:
> 
> * no distance and clean:
>     `{tag}`
> * distance and clean:
>     `{next_version}.dev{distance}+{scm letter}{revision hash}`
> * no distance and not clean:
>     `{tag}+dYYYYMMDD`
> * distance and not clean:
>     `{next_version}.dev{distance}+{scm letter}{revision hash}.dYYYYMMDD`
> 
> The next version is calculated by adding 1 to the last numeric component of the tag.

I presumed we were the "distance and clean" camp, hence:

* `{next_version}.dev{distance}+{scm letter}{revision hash}`, i.e.
* `2.0.4.dev0+g5929d76.d20220603`

It might be possible to change the default configuration to not automatically go for `.dev`?

kousu replied:
Why are we generating an extra commit? 

We should be able to use it like

* https://github.com/kousu/radicale-bsdauth/actions/runs/2203135595
* https://github.com/kousu/radicale-bsdauth/actions/runs/2203464045

btw, in those repos I eventually decided to drop workflow_dispatch entirely: you just click the New Release button, github tags it, the build is triggered. If you want to do a changelog you just make sure its committed to master first.

joshuacwnewton replied:
> But I really think we should use it, it really improves the workflow.

I don't quite understand the benefit of using `setuptools_scm` to fetch existing tags when the person making the releases must explicitly specify the version number anyway. Could you explain your thought process here?

Using environment variables (like was done in [804b510](https://github.com/spinalcordtoolbox/data-template/pull/1/commits/804b5104e56347ff1cd7f35fdde809824f6b1cdb)) seems straightforward enough to me? 🤔

joshuacwnewton replied:
> Why are we generating an extra commit?

We aren't?

kousu replied:
...yeah right, I've read the whole thread now and the problem is that workflow_dispatch doesn't create the tag until the end and that complicates things. It's got a couple other difficult corner cases and it's just not worth trying to deal with imo!

joshuacwnewton replied:
Ah, I see!! I think I get why it's worked for you.

So, with `setuptools_scm`, if you go the `release` route, then you end up in the ["no distance and clean" case](https://github.com/spinalcordtoolbox/data-template/pull/1#discussion_r889239510), at which point `{tag}` is used straight-up. 

(In other words, they only append `.dev` if you're trying to release from an untagged commit.)

joshuacwnewton replied:
I agree with you that we should drop `workflow_dispatch`. Much nicer to just have a single "correct" way of doing things, and the release UI has a lot of other nice usage benefits, too (e.g. checking if you're duplicating a tag name).

kousu replied:
Ai yi yi. Glad we cleared that up. Sorry if I rushed to answer too fast!

joshuacwnewton replied:
> Ah, I see!! I think I get why it's worked for you.
> 
> So, with `setuptools_scm`, if you go the `release` route, then you end up in the ["no distance and clean" case](https://github.com/spinalcordtoolbox/data-template/pull/1#discussion_r889239510), at which point `{tag}` is used straight-up.
> 
> (In other words, they only append `.dev` if you're trying to release from an untagged commit.)

I thought we had cleared this up. But, I'm still running into the old `.dev0` automatic version bumping with `setuptools_scm`.

* I forked `data-template` to my personal GitHub ccount.
* I merged the PR as-is (for the purposes of rapid testing in my fork on `trunk`).
* I [re-added `setuptools_scm`](https://github.com/joshuacwnewton/data-template/commit/e13b4f8673a3ddb338012191ee63cd016a791220).
* I then made a release, which [triggered the actions workflow](https://github.com/joshuacwnewton/data-template/runs/6762750946?check_suite_focus=true).
   * The commit is: https://github.com/joshuacwnewton/data-template/commit/e13b4f8673a3ddb338012191ee63cd016a791220
   * The version tag is: https://github.com/joshuacwnewton/data-template/tree/2022.6.6, which corresponds to the above commit.
   * In the workflow, we can confirm the correct commit is checked out [here](https://github.com/joshuacwnewton/data-template/runs/6762750946?check_suite_focus=true#step:3:108)
   * Yet, for some reason, instead of `2022.6.6`, the build command uses [`2022.6.7.dev0+ge13b4f8.d20220606`](https://github.com/joshuacwnewton/data-template/runs/6762750946?check_suite_focus=true#step:7:22)

I don't really know why my `setuptools_scm` setup is behaving differently compared to https://github.com/spinalcordtoolbox/data-template/pull/1#discussion_r889243135. Is it because we're not using a `pyproject.toml` file? Is the version of `setuptools_scm` buggy/outdated? 

I'm going to try and investigate this more, but my first impressions of `setuptools_scm` have been that it's an overly complicated solution for the simple versioning use-case we're trying to go for here (as opposed to just `perl -pi -e`'ing the version number directly into `setup.py`). 

I'm still open to being convinced, but, grappling with this has been a little vexing. :P

joshuacwnewton replied:
Hmm. Is this due to the fact that `actions/checkout` checks out the **commit**, and not the **tag**? (Hence the `.dev0`: We're on a git ref that's `0` distance away from the most recent tag, but we haven't actually checked out the tag itself.)

EDIT: Nope. [Checking out the tag itself](https://github.com/joshuacwnewton/data-template/runs/6763017479?check_suite_focus=true#step:3:88) still generates the pip version [`2022.6.7.dev0+ge13b4f8.d20220606`](https://github.com/joshuacwnewton/data-template/runs/6763017479?check_suite_focus=true#step:7:22).

EDIT2: I tried bumping the versions of the checkout action just in case that was having some sort of effect, but nope. 

The only differences I can see between this project and `radicale-bsdauth` are:

* `radicale-bsdauth` uses `pyproject.toml`, this uses `setup.py`.
* `radicale-bsdauth` uses `rc1`-based version numbering, while `setup.py` uses "final release" numbering.

joshuacwnewton replied:
I'm going to keep the `perl -pi -e` form of versioning for now just so I can rapidly prototype an SCT pip install in my own forks.

But, if you have any thoughts for why this might be happening, please feel free to chime in, @kousu.

kousu replied:
It's because of the `perl -pi`: you can't modify anything after the tag, if you do that's no longer the tag. **setuptools_scm and `python -m build` and the other works PyPA is trying is here to force us into these better practices** (though I still think pyproject.toml is dumb at the moment).

I tried building off your tag locally and instead of running `perl -pi` I made the equivalent edit, committed, and retagged it (pretending to be tomorrow), and that worked:

```
(a) p115628@joplin:~/src/data-template$ git log -p
commit 12aad1788069ba7ed277096cae3c9aad80524384 (HEAD, tag: 2022.6.7)
Author: Nick Guenther <nick.guenther@polymtl.ca>
Date:   Mon Jun 6 17:20:16 2022 -0400

    change dataset name

diff --git a/setup.py b/setup.py
index beebdc1..235e940 100644
--- a/setup.py
+++ b/setup.py
@@ -8,7 +8,7 @@ here = pathlib.Path(__file__).parent.resolve()
 import site; site.ENABLE_USER_SITE = True
 
 setup(
-  name='spinalcordtoolbox-data-${dataset_name}',
+  name='spinalcordtoolbox-data-t1',
   version="dev",
   description='Part of https://github.com/neuropoly/spinalcordtoolbox',
   long_description=(here / 'README.md').read_text(encoding='utf-8'),

commit e13b4f8673a3ddb338012191ee63cd016a791220 (tag: 2022.6.6)
Author: Joshua Newton <joshuacwnewton@gmail.com>
Date:   Mon Jun 6 16:29:53 2022 -0400

    `setup.py`: Re-add `setuptools_scm`
```

<details><summary>Successfully built spinalcordtoolbox-data-t1-2022.6.7.tar.gz and spinalcordtoolbox_data_t1-2022.6.7-py3-none-any.whl</summary>

```
(a) p115628@joplin:~/src/data-template$ python -m build --wheel --sdist
* Creating venv isolated environment...
* Installing packages in isolated environment... (setuptools>=40.8.0, setuptools_scm[toml], wheel)
* Getting dependencies for sdist...
* Installing packages in isolated environment... (setuptools, setuptools_scm[toml], wheel)
* Building sdist...
running sdist
running egg_info
writing src/spinalcordtoolbox_data_t1.egg-info/PKG-INFO
writing dependency_links to src/spinalcordtoolbox_data_t1.egg-info/dependency_links.txt
writing top-level names to src/spinalcordtoolbox_data_t1.egg-info/top_level.txt
package init file 'src/spinalcordtoolbox/__init__.py' not found (or not a regular file)
package init file 'src/spinalcordtoolbox/data/__init__.py' not found (or not a regular file)
adding license file 'LICENSE.txt'
writing manifest file 'src/spinalcordtoolbox_data_t1.egg-info/SOURCES.txt'
running check
creating spinalcordtoolbox-data-t1-2022.6.7
creating spinalcordtoolbox-data-t1-2022.6.7/.github
creating spinalcordtoolbox-data-t1-2022.6.7/.github/workflows
creating spinalcordtoolbox-data-t1-2022.6.7/src
creating spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox
creating spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox/data
creating spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox/data/dataset
creating spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
copying files to spinalcordtoolbox-data-t1-2022.6.7...
copying .data-template-new.png -> spinalcordtoolbox-data-t1-2022.6.7
copying .gitignore -> spinalcordtoolbox-data-t1-2022.6.7
copying LICENSE.txt -> spinalcordtoolbox-data-t1-2022.6.7
copying README.md -> spinalcordtoolbox-data-t1-2022.6.7
copying pyproject.toml -> spinalcordtoolbox-data-t1-2022.6.7
copying setup.py -> spinalcordtoolbox-data-t1-2022.6.7
copying .github/workflows/release.yml -> spinalcordtoolbox-data-t1-2022.6.7/.github/workflows
copying src/spinalcordtoolbox/data/dataset/__init__.py -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox/data/dataset
copying src/spinalcordtoolbox_data_t1.egg-info/PKG-INFO -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
copying src/spinalcordtoolbox_data_t1.egg-info/SOURCES.txt -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
copying src/spinalcordtoolbox_data_t1.egg-info/dependency_links.txt -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
copying src/spinalcordtoolbox_data_t1.egg-info/not-zip-safe -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
copying src/spinalcordtoolbox_data_t1.egg-info/top_level.txt -> spinalcordtoolbox-data-t1-2022.6.7/src/spinalcordtoolbox_data_t1.egg-info
Writing spinalcordtoolbox-data-t1-2022.6.7/setup.cfg
Creating tar archive
removing 'spinalcordtoolbox-data-t1-2022.6.7' (and everything under it)
* Creating venv isolated environment...
* Installing packages in isolated environment... (setuptools>=40.8.0, setuptools_scm[toml], wheel)
* Getting dependencies for wheel...
* Installing packages in isolated environment... (setuptools, setuptools_scm[toml], wheel)
* Building wheel...
running bdist_wheel
running build
running build_py<
package init file 'src/spinalcordtoolbox/__init__.py' not found (or not a regular file)
package init file 'src/spinalcordtoolbox/data/__init__.py' not found (or not a regular file)
running egg_info
writing src/spinalcordtoolbox_data_t1.egg-info/PKG-INFO
writing dependency_links to src/spinalcordtoolbox_data_t1.egg-info/dependency_links.txt
writing top-level names to src/spinalcordtoolbox_data_t1.egg-info/top_level.txt
adding license file 'LICENSE.txt'
writing manifest file 'src/spinalcordtoolbox_data_t1.egg-info/SOURCES.txt'
installing to build/bdist.linux-x86_64/wheel
running install
running install_lib
creating build/bdist.linux-x86_64/wheel
creating build/bdist.linux-x86_64/wheel/spinalcordtoolbox
creating build/bdist.linux-x86_64/wheel/spinalcordtoolbox/data
creating build/bdist.linux-x86_64/wheel/spinalcordtoolbox/data/dataset
copying build/lib/spinalcordtoolbox/data/dataset/__init__.py -> build/bdist.linux-x86_64/wheel/spinalcordtoolbox/data/dataset
running install_egg_info
Copying src/spinalcordtoolbox_data_t1.egg-info to build/bdist.linux-x86_64/wheel/spinalcordtoolbox_data_t1-2022.6.7-py3.10.egg-info
running install_scripts
adding license file "LICENSE.txt" (matched pattern "LICEN[CS]E*")
creating build/bdist.linux-x86_64/wheel/spinalcordtoolbox_data_t1-2022.6.7.dist-info/WHEEL
creating '/home/GRAMES.POLYMTL.CA/p115628/src/data-template/dist/tmphcd27lwe/spinalcordtoolbox_data_t1-2022.6.7-py3-none-any.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
adding 'spinalcordtoolbox/data/dataset/__init__.py'
adding 'spinalcordtoolbox_data_t1-2022.6.7.dist-info/LICENSE.txt'
adding 'spinalcordtoolbox_data_t1-2022.6.7.dist-info/METADATA'
adding 'spinalcordtoolbox_data_t1-2022.6.7.dist-info/WHEEL'
adding 'spinalcordtoolbox_data_t1-2022.6.7.dist-info/top_level.txt'
adding 'spinalcordtoolbox_data_t1-2022.6.7.dist-info/RECORD'
removing build/bdist.linux-x86_64/wheel
Successfully built spinalcordtoolbox-data-t1-2022.6.7.tar.gz and spinalcordtoolbox_data_t1-2022.6.7-py3-none-any.whl
```

</details>

kousu replied:
We just need to ask people need to manually fill in the name.

Maybe we can write some awesome actions script that edits setup.py to sync it with the repo's name on every commit. That would be cool, and in line with using git as the single source of truth, but maybe a bit Extra, y'know? I think doing step manually is good enough because how many times are we really going to be making a new dataset (vs making a new release of that dataset)?

joshuacwnewton replied:
Oh of course. I missed the fact that editing `setup.py` means that the git repo is no longer in a clean state. 

Thanks for catching that. :facepalm:

joshuacwnewton replied:
> We just need to ask people need to manually fill in the name.

We already do! The `perl` edit step is only relevant for the `data-template` repo, and I only added it because otherwise the build on `data-template` (i.e. this PR) would fail due to (I think) an invalid repo URL in the metadata, hehe.

We can just straight-up remove the `perl` step, and allow the builds to fail on `data-template`? Since the problem goes away the moment people actually use the template then follow the README.

joshuacwnewton replied:
From @kousu: 

> Ah good. I was pretty sure we already did...

joshuacwnewton replied:
I tried out the template using PAM50, and all the steps went well: https://github.com/joshuacwnewton/data-PAM50/releases/tag/2022.6.6

I then tried installing the wheel and it behaved as expected.

So, as soon as I add back `setuptools_scm`, this (maybe possibly) is almost ready for merging?

joshuacwnewton committed:
`release.yml`: Try tagging release prior to build

joshuacwnewton committed:
Revert "`release.yml`: Try tagging release prior to build"

This reverts commit 73f9b339b34197eed87796ffd0e67123bab3235c.

joshuacwnewton committed:
`release.yml`: Explicitly set `version=` instead of using setuptools_scm

joshuacwnewton committed:
`release.yml`: Make `VERSION_NUMBER` more explicit

* Stop repeating version number
* Use || to make fallback case more explicit

joshuacwnewton committed:
`release.yml`: Drop `workflow_dispatch`

It's more straightforward to just have a single "correct" way of doing things, and the release UI has a lot of other nice usage benefits, too (e.g. checking if you're duplicating a tag name).

joshuacwnewton committed:
`release.yml`: Drop test build workflow, and use only 1 workflow

The `startsWith(github.ref, 'refs/tags/')` pattern is recommended by:

* https://github.com/softprops/action-gh-release
* https://github.com/pypa/gh-action-pypi-publish

joshuacwnewton committed:
README.md: Fix typo (missing `)

joshuacwnewton commented on README.md:
```diff
@@ -0,0 +1,103 @@
+2. Go to https://github.com/spinalcordtoolbox/data-${dataset_name}/releases
+
+    1. Click "Draft Release"
+    2. Fill in a version tag. We use date-based release names (e.g. `r20200101`, `r20220518`, etc.)
```
Just realizing now... date-based release names aren't kosher with PyPI, are they, since PyPI expects SemVer? Or is that true?

I feel like I need to try this on test.pypi.org out before being able to comfortable recommend this.

kousu replied:
Actually they are. One of the reasons https://peps.python.org/pep-0440/ is so complicated is because it didn't want to offend projects using date-based or even [pi-based](https://en.wikipedia.org/wiki/TeX#TeX82) or even the time-based choice [pytz uses](https://peps.python.org/pep-0440/#olson-database-versioning).

kousu replied:
![Screenshot 2022-06-03 at 20-39-32 PEP 440 – Version Identification and Dependency Specification peps python org](https://user-images.githubusercontent.com/987487/171969678-108ff15a-b8ba-426a-8bb7-cb3074c69281.png)

(dropping the dots between year and month still works right, the relative ordering is the same)

kousu replied:
Oh one thing though is the 'r' might not be kosher. It [should be at most a v](https://peps.python.org/pep-0440/#preceding-v-character)

joshuacwnewton replied:
Thank you!! :)

joshuacwnewton replied:
Yep, all good: https://test.pypi.org/project/spinalcordtoolbox-data-template/2022.6.3/

joshuacwnewton committed:
`release.yml`: Test version tag `r20220603`

joshuacwnewton committed:
`release.yml`: Test version tag `2022.06.03`

joshuacwnewton committed:
`release.yml`: Use quotes for version string

joshuacwnewton committed:
`release.yml`: Temporarily stop using `||`

I just want to see how pip will use date-based versions! GitHub Actions, pls stop being so finicky.

joshuacwnewton committed:
`README.md`: Fix typo (missing `)

joshuacwnewton committed:
`release.yml`: Revert back to using release tag

joshuacwnewton committed:
`release.yml`: Re-enable 'publish' if statements

joshuacwnewton committed:
`README.md`: Clarify date-based version numbering

kousu commented:
Ah good. I was pretty sure we already did...
